### ISSUE TITLE
couldn't execute "/u01/app/oracle/product/12/OPatch/ocm/bin/emocmrsp": permission denied

### ISSUE THREAD

**Issue by goutham27** [biemond/biemond-oradb#111]
Hi,

I am getting below permission denied error when i try to do opatch upgrade, I was able to fix this by adding a file resource to change file permission of emocmrsp file. Could you pls check and let me know..If i am doing something wrong.

```
==> ve_oracle_centos6: Debug: Exec[extract opatch opatch_upgrade opatch.zip](provider=posix): Executing 'unzip -o /var/tmp/install/opatch.zip -d /u01/app/oracle/product/12'
==> ve_oracle_centos6: Debug: Executing with uid=oracle gid=oinstall: 'unzip -o /var/tmp/install/opatch.zip -d /u01/app/oracle/product/12'
==> ve_oracle_centos6: Notice: /Stage[main]/Ve_oracle::Patch/Oradb::Opatchupgrade[opatch_upgrade]/Exec[extract opatch opatch_upgrade opatch.zip]/returns: executed successfully
==> ve_oracle_centos6: Debug: /Stage[main]/Ve_oracle::Patch/Oradb::Opatchupgrade[opatch_upgrade]/Exec[extract opatch opatch_upgrade opatch.zip]: The container Oradb::Opatchupgrade[opatch_upgrade] will propagate my refresh event
==> ve_oracle_centos6: Debug: Exec[ksh /var/tmp/install/opatch_upgrade_opatch_upgrade_12.1.0.1.9.ksh](provider=posix): Executing 'ksh /var/tmp/install/opatch_upgrade_opatch_upgrade_12.1.0.1.9.ksh'
==> ve_oracle_centos6: Debug: Executing with uid=oracle gid=oinstall: 'ksh /var/tmp/install/opatch_upgrade_opatch_upgrade_12.1.0.1.9.ksh'
==> ve_oracle_centos6: Notice: /Stage[main]/Ve_oracle::Patch/Oradb::Opatchupgrade[opatch_upgrade]/Exec[ksh /var/tmp/install/opatch_upgrade_opatch_upgrade_12.1.0.1.9.ksh]/returns: spawn /u01/app/oracle/product/12/OPatch/ocm/bin/emocmrsp -no_banner
==> ve_oracle_centos6: Notice: /Stage[main]/Ve_oracle::Patch/Oradb::Opatchupgrade[opatch_upgrade]/Exec[ksh /var/tmp/install/opatch_upgrade_opatch_upgrade_12.1.0.1.9.ksh]/returns: couldn't execute "/u01/app/oracle/product/12/OPatch/ocm/bin/emocmrsp": permission denied
==> ve_oracle_centos6: Notice: /Stage[main]/Ve_oracle::Patch/Oradb::Opatchupgrade[opatch_upgrade]/Exec[ksh /var/tmp/install/opatch_upgrade_opatch_upgrade_12.1.0.1.9.ksh]/returns:     while executing
==> ve_oracle_centos6: Notice: /Stage[main]/Ve_oracle::Patch/Oradb::Opatchupgrade[opatch_upgrade]/Exec[ksh /var/tmp/install/opatch_upgrade_opatch_upgrade_12.1.0.1.9.ksh]/returns: "spawn /u01/app/oracle/product/12/OPatch/ocm/bin/emocmrsp -no_banner"
==> ve_oracle_centos6: Debug: Oradb::Opatchupgrade[opatch_upgrade]: Resource is being skipped, unscheduling all events
==> ve_oracle_centos6: Info: Oradb::Opatchupgrade[opatch_upgrade]: Unscheduling all events on Oradb::Opatchupgrade[opatch_upgrade]
==> ve_oracle_centos6: Notice: /Stage[main]/Ve_oracle::Patch/Exec[download-patch]: Dependency Exec[ksh /var/tmp/install/opatch_upgrade_opatch_upgrade_12.1.0.1.9.ksh] has failures: true
==> ve_oracle_centos6: Debug: /Stage[main]/Ve_oracle::Patch/Exec[download-patch]: Resource is being skipped, unscheduling all events
==> ve_oracle_centos6: Debug: Oradb::Opatch[_db_patch]: Resource is being skipped, unscheduling all events
==> ve_oracle_centos6: Notice: /Stage[main]/Ve_oracle::Patch/Oradb::Opatch[_db_patch]/File[/var/tmp/install/p22191659_121020_Linux-x86-64.zip]: Dependency Exec[ksh /var/tmp/install/opatch_upgrade_opatch_upgrade_12.1.0.1.9.ksh] has failures: true
==> ve_oracle_centos6: Debug: /Stage[main]/Ve_oracle::Patch/Oradb::Opatch[_db_patch]/File[/var/tmp/install/p22191659_121020_Linux-x86-64.zip]: Resource is being skipped, unscheduling all events
==> ve_oracle_centos6: Notice: /Stage[main]/Ve_oracle::Patch/Oradb::Opatch[_db_patch]/Exec[extract opatch p22191659_121020_Linux-x86-64.zip _db_patch]: Dependency Exec[ksh /var/tmp/install/opatch_upgrade_opatch_upgrade_12.1.0.1.9.ksh] has failures: true
==> ve_oracle_centos6: Debug: /Stage[main]/Ve_oracle::Patch/Oradb::Opatch[_db_patch]/Exec[extract opatch p22191659_121020_Linux-x86-64.zip _db_patch]: Resource is being skipped, unscheduling all events
==> ve_oracle_centos6: Error: ksh /var/tmp/install/opatch_upgrade_opatch_upgrade_12.1.0.1.9.ksh returned 1 instead of one of [0]
==> ve_oracle_centos6: Error: /Stage[main]/Ve_oracle::Patch/Oradb::Opatchupgrade[opatch_upgrade]/Exec[ksh /var/tmp/install/opatch_upgrade_opatch_upgrade_12.1.0.1.9.ksh]/returns: change from notrun to 0 failed: ksh /var/tmp/install/opatch_upgrade_opatch_upgrade_12.1.0.1.9.ksh returned 1 instead of one of [0]
```

**Below is my code**

```
oradb::opatchupgrade{'opatch_upgrade':
  oracle_home               => hiera('oracle_home_dir'),
  patch_file                => 'opatch.zip',
  opversion                 => '12.1.0.1.9',
  user                      => hiera('oracle_os_user'),
  group                     => 'oinstall',
  download_dir              => hiera('oracle_download_dir'),
  puppet_download_mnt_point => hiera('oracle_source'),
} 

file { '/u01/app/oracle/product/12/OPatch/ocm/bin/emocmrsp':
   ensure => present,
   mode   => '0755',
   owner  => 'oracle',
   group  => 'oinstall',
}  
```

**Comment by biemond** [comment]
Hi,

I don't know. What are the permission of this file and did you install Oracle home under a different user/group. Or maybe it is the umask

Thanks

**Comment by goutham27** [comment]
Hi,

I have used oracle user and dba group to install, umask is set to 0022. Also I had to change the opatch executable permission, otherwise I am getting a permission denied while installing the patches. 

oracle_home_dir:          '/u01/app/oracle/product/12'
oracle_os_user:           'oracle'
oracle_os_group:          'dba'

Only thing I am doing different is downloading the opatch from http source to /tmp directory.

    exec { 'download_opatch':
      command   => "curl $download_opatch -o /tmp/opatch.zip",
      path      => '/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/sbin',
      logoutput => true,
    }

 Are there any plans to support adding http source to download database/opatch etc..?
